### PR TITLE
usb: redesign of  USB transfers

### PR DESCRIPTION
**Pull request by jfischer-no**
Redesign of USB transfers.

This is draft for rework on USB transfers.
Split off #21350 and still based on it.

### CODE REVIEW

**Review by eobalski — COMMENTED** [review]
Reading this transfer mechanism is making me anxious. I got the basic idea, but I am not sure if its good that we want to make every transfer rely on 'usb transfers'. For instance #21526 is using ISO endpoints which transfer a lot of data. Adding additional mechanism seems unnecessary for me (as for now, but I may be missing something).

If We want to unify everything I think it is worth to discuss how should we deal with different transfers (ISO in particular).

For now I've tested it on several samples and for eg. cdc_acm_composite has problem to get into configured state (finally it reaches it) but then there is problem to open a console for testing (minicom used). I have not analyzed the exact cause. I just saw some transfer errors from usb_transfer.
Once I managed to make it work, but it behaves randomly.

**Comment by jfischer-phytec-iot** [comment]
> Reading this transfer mechanism is making me anxious. I got the basic idea, but I am not sure if its good that we want to make every transfer rely on 'usb transfers'. For instance #21526 is using ISO endpoints which transfer a lot of data. Adding additional mechanism seems unnecessary for me (as for now, but I may be missing something).
> 
> If We want to unify everything I think it is worth to discuss how should we deal with different transfers (ISO in particular).

usb_transfer is not really usable for use cases other than netusb, this PR is for testing. a) we need a unified interface for the transfer, even for ISO transfers. b) for many driver in the tree endpoint callback are executed in ISR context, following control request (also class and vendor request) are handled in the ISR context. This PR resolves this issue. Also packet refragmentation will be handled by usb_transfer and we can simplify the code.

> 
> For now I've tested it on several samples and for eg. cdc_acm_composite has problem to get into configured state (finally it reaches it) but then there is problem to open a console for testing (minicom used). I have not analyzed the exact cause. I just saw some transfer errors from usb_transfer.
> Once I managed to make it work, but it behaves randomly.

That should not happen, at least increasing MAX_NUM_TRANSFERS have to fix it.

File: subsys/usb/usb_transfer.c [hunk 1]
```diff
@@ -159,12 +162,8 @@ void usb_transfer_ep_callback(u8_t ep, enum usb_dc_ep_cb_status_code status)
 {
 	struct usb_transfer_data *trans = usb_ep_get_transfer(ep);
 
-	if (status != USB_DC_EP_DATA_IN && status != USB_DC_EP_DATA_OUT) {
-		return;
-	}
-
 	if (!trans) {
-		if (status == USB_DC_EP_DATA_OUT) {
+		if (status == USB_DC_EP_DATA_OUT || status == USB_DC_EP_SETUP) {
```
Comment (emob-nordic) [inline subsys/usb/usb_transfer.c]: Can we concatenate those 2 ifs?

File: subsys/usb/usb_transfer.c [hunk 2]
```diff
@@ -181,6 +180,10 @@ void usb_transfer_ep_callback(u8_t ep, enum usb_dc_ep_cb_status_code status)
 			LOG_ERR("RX data lost, no transfer");
 		}
 		return;
+	} else {
+		if ((ep == USB_CONTROL_OUT_EP0 || ep == USB_CONTROL_IN_EP0)) {
+			control_ep_status = status;
+		}
 	}
 
 	if (!k_is_in_isr() || (status == USB_DC_EP_DATA_OUT)) {
```
Comment (emob-nordic) [inline subsys/usb/usb_transfer.c]: As I understand this will not be needed anymore, should ``` else ``` be deleted?

**Review by eobalski — COMMENTED** [review]
I like the idea of unifying transfers, it would be great to have this.
I have some concerns:

There is a need to manually kick the whole machinery for OUT transfers. For what I saw it is impossible to avoid it if we decide to follow transfer mechanism. This is already done for several classes ([cdc](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/cdc_acm.c#L306), [bluetooth](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/bluetooth.c#L206)).
	
From stack perspective classes should not bother about any transfers (right now they do when CONFIGURED state reached) . The whole 'endpoint related stuff' should be left apart from what classes exposes to the user (this is my opinion).
	
To satisfy this i think we can try to start USB OUT transfers inside usb_enable, similar to what You have done [here](https://github.com/zephyrproject-rtos/zephyr/pull/21490/files#diff-609952c961f3c77cb7811bedee4bfd2cR1277) by doing something like:
```
for reach(usb_cfg_data)
   if(endpoint == OUT)
      usb_transfer()
```

Something similar to [foreach_ep](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L937).

This will require reimplementation of buffer management.
Maybe it is worth think about it as a whole. Let me know what You think.

**Comment by jfischer-phytec-iot** [comment]
close in favor of #27790

**Pull request closed without merging**